File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    const map = new Map();
    const days = 10;
    const wantSize = want.length;
    let answer = 0;

    for (let i = 0; i < days; i++) {
        map.set(discount[i], (map.get(discount[i]) ?? 0) + 1);
    }

    function isValid() {
        for (let i = 0; i < wantSize; i++) {
            if ((map.get(want[i]) ?? 0) < number[i]) return false;
        }
        return true;
    }

    if (isValid()) answer++;

    for (let i = days; i < discount.length; i++) {
        let removeItem = discount[i - days];
        let cnt = map.get(discount[i-days]) - 1;
        if(cnt == 0){
            map.delete(discount[i-days]);
        }else{
            map.set(discount[i-days], cnt);
        }

        let newItem = discount[i];
        map.set(newItem, (map.get(newItem) ?? 0) + 1);

        if (isValid()) answer++;
    }

    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public String[] want;
    public int[] number;
    public HashMap<String, Integer> map = new HashMap<>();
    
    public int solution(String[] want, int[] number, String[] discount) {
        this.want = want;
        this.number = number;
        int minDay = 10;
        int answer = 0;
        
        for(int i = 0; i < discount.length; i++){
            map.put(discount[i], map.getOrDefault(discount[i], 0) + 1);
            if(i >= 9){
                if(isValid()){
                    answer++;
                }
                map.put(discount[i - 9], map.get(discount[i - 9]) - 1);
            }
        }
        return answer;
    }
    
    public boolean isValid(){
        for(int i = 0; i < want.length; i++){
            if(map.getOrDefault(want[i], 0) < number[i]){
                return false;
            }
        }
        return true;
    }
}